proposals/init.md

```diff
@@ -0,0 +1,582 @@
+    {
+        Property = 42;  // Okay 
+        Field = 13;     // Error Field is readonly
+    }
+}
+```
+
+When `init` is used in a virtual property then all the overrides must also
+be marked as `init`. Likewise it is not possible to override a simple 
+`set` with `init`.
+
+```cs
+class Base
+{
+    public virtual int Property { get; init; }
+}
+
+class C1 : Base
+{
+    public override int Property { get; init; }
+}
+
+class C2 : Base
+{
+    // Error: Property must have init to override Base.Property
+    public override int Property { get; set; }
+}
+```
+
+An `interface` declaration can also particpate in `init` style initalization 
+via the following pattern:
+
+```cs
+interface IPerson
+{
+    string Name { get; init; }
+}
+
+class Init
+{
+    void M<T>() where T : IPerson, new()
+    {
+        var local = new T()
+        {
+            Name = "Jared"
+        };
+        local.Name = "Jraed"; // Error
+    }
+}
+```
+
+Restrictions of this feature:
+- The `init` accessor can only be used on instance properties
+- A property cannot contain both an `init` and `set` accessor
+- All overrides of a property must have `init` if the base had `init`. This rule
+also applies to interface implementation.
+
+### Metadata encoding 
+Property `init` accessors will be emitted as a standard `set` accessor with
+the return type marked with a modreq of `IsInitOnly`. This is a new type
+which will have the following definition:
+
+```cs
+namespace System.Runtime.CompilerServices
+{
+    public sealed class IsInitOnly
+    {
+
+    }
+}
+```
+
+The compiler will match the type by full name. There is no requirement
+that it appear in the core library. If there are mulitple types by this name
+then the compiler will tie break in the following order:
+
+1. The one defined in the project being compiled
+1. The one defined in corelib
+
+If neither of these exist then a type ambiguity error will be issued.
+
+The design for `IsInitOnly` is futher covered in [this issue](https://github.com/dotnet/runtime/issues/34978)
+
+## Questions
+
+### Breaking changes
+One of the main pivot points in how this feature is encoded will come down to
+the following question: 
+
+> Is it a binary breaking change to replace `init` with `set`?
+
+Replacing `init` with `set` and thus making a property fully writable is never
+a source breaking change on a non-virtual property. It simply expands the set
+of scenarios where the property can be written. The only behavior in question is
+whether or not this remains a binary breaking change.
+
+If we want to make the change of `init` to `set` a source and binary compatible
+change then it will force our hand on the modreq vs. attributes decision
+below because it will rule out modreqs as a soulution. If on the other hand
+this is seen as a non-interesting then this will make the modreq vs. attribute
+decision less impactful.
+
+**Resolution**
+This scenario is not seen as compelling by LDM.
+
+### Modreqs vs. attributes
```

Comment:
Do you have a link that explains what a modreq is? This is a new term for me personally.

Reply:
Yeah it's a term very few understand. Probably only people who've had to deal with metadata encoding / decoding.

https://twitter.com/jaredpar/status/1247014490069020672

I don't have a good link to a document describing what a modreq or modopt is other than some obscure passages in the ECMA CLI document [(here)](https://twitter.com/ben_a_adams/status/1247016887600533505). Should probably write something up about it at some point. 

Probably the easiest way to think about them is they're like attributes on method signatures but they're actually a part of the signature. That its an attribute that must be repeated in the same position whenever a method is overridden or when a method implements an interface members. They have the further requirement than if a compiler doesn't understand the type inside a modreq then it should effectively ignore the containing member altogether.

Reply:
I see, thanks for the explanation! So basically custom markers for the compiler and the runtime to handle stuff in a special way. Works for me :D
One final question: What does the “mod” in “modreq” and “modopt” stand for? Req is likely requirement and opt is option, but mod? “Modification” as in to make a modification on the generated IL?

Reply:
"Mod" stands for modifier.

Reply:
More specifically:

- modreq = modifier required: that is it is required that the compiler reading the entry understand the type in the modreq. That is because the type in the modreq implies special behavior that compilers are required to understand in order to handle the function correctly. Hence if the compiler does not understand it should discard that member.
- modopt = modifier optional: that is the compiler is not required to understand the type in the modopt. The type indicates behavior that can be beneficial to compilers who understand the behavior but knowledge is not required in order to correctly invoke the member.